src/engine/DnCWorker.cpp

```diff
@@ -60,132 +60,122 @@ void DnCWorker::setQueryDivider( DivideStrategy divideStrategy )
-                delete subQuery;
-            }
-            else if ( result == Engine::TIMEOUT )
+        // Apply the split and solve
+        _engine->applySplit( *split );
+        _engine->solve( timeoutInSeconds );
+
+        IEngine::ExitCode result = _engine->getExitCode();
+        printProgress( queryId, result );
+        // Switch on the result
+        if ( result == IEngine::UNSAT )
+        {
+            // If UNSAT, continue to solve
+            *_numUnsolvedSubQueries -= 1;
+            if ( _numUnsolvedSubQueries->load() == 0 )
+                *_shouldQuitSolving = true;
+            delete subQuery;
+        }
+        else if ( result == IEngine::TIMEOUT )
+        {
+            // If TIMEOUT, split the current input region and add the
+            // new subQueries to the current queue
+            SubQueries subQueries;
+            _queryDivider->createSubQueries( pow( 2, _onlineDivides ),
+                                             queryId, *split,
+                                             (unsigned) timeoutInSeconds *
+                                             _timeoutFactor, subQueries );
+            for ( auto &newSubQuery : subQueries )
             {
-                // If TIMEOUT, split the current input region and add the
-                // new subQueries to the current queue
-                SubQueries subQueries;
-                _queryDivider->createSubQueries( pow( 2, _onlineDivides ),
-                                                 queryId, *split,
-                                                 (unsigned) timeoutInSeconds *
-                                                 _timeoutFactor, subQueries );
-                for ( auto &newSubQuery : subQueries )
+                if ( !_workload->push( std::move( newSubQuery ) ) )
                 {
-                    if ( !_workload->push( std::move( newSubQuery ) ) )
-                    {
-                        ASSERT( false );
-                    }
-
-                    *_numUnsolvedSubQueries += 1;
+                    ASSERT( false );
                 }
-                *_numUnsolvedSubQueries -= 1;
-                delete subQuery;
+
+                *_numUnsolvedSubQueries += 1;
+            }
+            *_numUnsolvedSubQueries -= 1;
+            delete subQuery;
+        }
+        else if ( result == IEngine::QUIT_REQUESTED )
+        {
+            // If engine was asked to quit, quit
+            std::cout << "Quit requested by manager!" << std::endl;
+            delete subQuery;
+            ASSERT( _shouldQuitSolving->load() );
             }
-            else if ( result == Engine::SAT )
+        else
+        {
+            // We must quit solving if the result is not UNSAT or TIMEOUT
+            *_shouldQuitSolving = true;
+            if ( result == IEngine::SAT )
             {
                 // If SAT, set the shouldQuitSolving flag to true, so that the
                 // DnCManager will kill all the DnCWorkers
```

Comment:
Comment says we should set the quit flag, but do we do that?

Reply:
We set the flag before case SAT/ERROR/NOT_DONE.
The comment in the branches might be misleading so I've updated them.